File: src/creature.js

import { NodeQueue } from './queue-data.js'
import { toKey } from './cave.js'

const MOVEMENT_OFFSET_MIN_DISTANCE = 1
const MOVEMENT_OFFSET_MAX_DISTANCE = 15

function getRandomMovementOffset(minDistance = MOVEMENT_OFFSET_MIN_DISTANCE, maxDistance = MOVEMENT_OFFSET_MAX_DISTANCE) {
    const safeMin = Number.isFinite(minDistance) ? minDistance : MOVEMENT_OFFSET_MIN_DISTANCE
    const safeMax = Number.isFinite(maxDistance) ? maxDistance : MOVEMENT_OFFSET_MAX_DISTANCE
    const clampedMax = Math.max(safeMin, safeMax)
    const angleDegrees = Math.random() * 360
    const angleRadians = angleDegrees * (Math.PI / 180)
    const distance = safeMin + (Math.random() * (clampedMax - safeMin))

    return {
        x: Math.cos(angleRadians) * distance,
        y: Math.sin(angleRadians) * distance
    }
}

export class Creature {
    constructor(name,location,sprite,game) {
        this.name = name
        this.queue = new NodeQueue()
        this.pathPreview = []
        this.health = 20
        this.maxHealth = this.health
        this.damage = 5
        this.location = location
        this.sprite = sprite
        this.game = game
        this.cave = null
        
        this.assignment = "unassigned"

        sprite.on('mouseup', (interactionEvent) => {
            if (this.game.buildMode || this.game.dragging) {
                return
            }
            
            if (this.game.selected.object === this) {
                this.game.selected.setSelected(null)
                return
            }

            if (this.game.selected.object) {
                this.game.selected.setSelected(null)
            }
            this.game.selected.setSelected(this)
            this.game.movePath = false
        })
    }

    clearActionQueue() {
        this.queue.clear()
        this.pathPreview = []
    }

    restartBehavior({ clearQueue = true } = {}) {
        if (clearQueue) {
            this.clearActionQueue()
        }

        if (typeof this.getBehavior !== 'function') {
            return false
        }

        const behavior = this.getBehavior()
        if (typeof behavior !== 'function') {
            return false
        }

        return behavior.call(this)
    }

    getDamage() {
        return this.damage
    }

    getMaxHealth() {
        return this.maxHealth
    }

    restoreHealth() {
        this.health = this.getMaxHealth()
        return this.health
    }

    dealDamage(target) {
        if (!target || target === this || typeof target.takeDamage !== 'function') {
            return 0
        }

        return target.takeDamage(this.getDamage(), this)
    }

    takeDamage(amount, source = null) {
        if (!Number.isFinite(amount) || amount <= 0 || this.health <= 0) {
            return 0
        }

        const applied = Math.min(this.health, amount)
        this.health -= applied

        if (this.health <= 0) {
            this.health = 0
            this.removeFromGame(source)
        }

        return applied
    }

    cleanupBeforeRemoval() {
        return
    }

    removeFromGame(source = null) {
        if (this.cave && typeof this.cave.removeCreature === 'function') {
            return this.cave.removeCreature(this, source)
        }

        this.clearActionQueue()
        this.cleanupBeforeRemoval()

        if (this.sprite?.parent) {
            this.sprite.parent.removeChild(this.sprite)
        }

        if (typeof this.sprite?.destroy === 'function') {
            this.sprite.destroy()
        }

        this.location = { x: null, y: null }
        this.cave = null
        return true
    }

    enqueueAction(actionFn) {
        if (typeof actionFn !== 'function') {
            return false
        }
        this.queue.enqueue(actionFn)
        return true
    }

    getNavigationFallback() {
        if (this.assignment === "miner" && typeof this.minerStep1 === 'function') {
            return () => this.minerStep1()
        }
        if (this.assignment === "farmer" && typeof this.farmerStep1 === 'function') {
            return () => this.farmerStep1()
        }
        if (this.assignment === "builder" && typeof this.builderStep1 === 'function') {
            return () => this.builderStep1()
        }
        if (this.assignment === "fighter" && typeof this.fighterStep1 === 'function') {
            return () => this.fighterStep1()
        }
        if (this.assignment === "enemy" && typeof this.enemyStep1 === 'function') {
            return () => this.enemyStep1()
        }
        return null
    }

    runNavigationFallback(fallbackFn) {
        this.clearActionQueue()
        if (typeof fallbackFn === 'function') {
            this.enqueueAction(() => fallbackFn.call(this))
        }
        return false
    }

    buildNavigationPathToPoint(destination) {
        if (!destination || !this.cave) {
            return null
        }

        const field = this.cave.buildPointBfsField(destination)
        if (!(field instanceof Map)) {
            return null
        }

        return this.cave.buildPathFromField(field, this.location)
    }

    buildNavigationPathToBuilding(building) {
        if (!building || !this.cave) {
            return null
        }

        const field = this.cave.ensureBuildingBfsField(building)
        if (!(field instanceof Map)) {
            return null
        }

        return this.cave.buildPathFromField(field, this.location)
    }

    recoverNavigation(destination, fallbackFn) {
        this.clearActionQueue()

        if (!destination) {
            return this.runNavigationFallback(fallbackFn)
        }

        const reroute = this.buildNavigationPathToPoint(destination)
        if (reroute && reroute.length > 1) {
            const target = { x: destination.x, y: destination.y }
            this._enqueueResolvedPath(reroute, () => this.recoverNavigation(target, fallbackFn), false)
            return false
        }

        if (reroute && reroute.length === 1) {
            return true
        }

        return this.runNavigationFallback(fallbackFn)
    }

    recoverBuildingNavigation(building, fallbackFn) {
        this.clearActionQueue()

        if (!building) {
            return this.runNavigationFallback(fallbackFn)
        }

        const reroute = this.buildNavigationPathToBuilding(building)
        if (reroute && reroute.length > 1) {
            this._enqueueResolvedPath(reroute, () => this.recoverBuildingNavigation(building, fallbackFn), false)
            return false
        }

        if (reroute && reroute.length === 1) {
            return true
        }

        return this.runNavigationFallback(fallbackFn)
    }

    executeNavigationStep(next, onFailure) {
        const result = this.cave.moveCreature(this, next)
        if (result === false) {
            return typeof onFailure === 'function' ? onFailure() : false
        }

        if (this.pathPreview.length > 0) {
            this.pathPreview.shift()
        }
        return result
    }

    _enqueueResolvedPath(path, onFailure, clearExisting) {
        if (clearExisting) {
            this.clearActionQueue()
        }
        if (!path || path.length < 2) {
            return false
        }
        const steps = [...path]
        steps.shift()

        for (const step of steps) {
            const next = { x: step.x, y: step.y }
            this.pathPreview.push(next)

            this.enqueueAction(() => {
                return this.executeNavigationStep(next, onFailure)
            })
        }

        return true
    }

    navigateTo(destination, fallbackFn = this.getNavigationFallback(), clearExisting = true) {
        if (!destination) {
            return this.runNavigationFallback(fallbackFn)
        }

        const path = this.buildNavigationPathToPoint(destination)
        if (!path) {
            return this.runNavigationFallback(fallbackFn)
        }
        if (path.length < 2) {
            return true
        }

        const target = { x: destination.x, y: destination.y }
        return this._enqueueResolvedPath(path, () => this.recoverNavigation(target, fallbackFn), clearExisting)
    }

    navigateToBuilding(building, fallbackFn = this.getNavigationFallback(), clearExisting = true) {
        if (!building) {
            return this.runNavigationFallback(fallbackFn)
        }

        const path = this.buildNavigationPathToBuilding(building)
        if (!path) {
            return this.runNavigationFallback(fallbackFn)
        }
        if (path.length < 2) {
            return true
        }

        return this._enqueueResolvedPath(path, () => this.recoverBuildingNavigation(building, fallbackFn), clearExisting)
    }

    queueMovePath(path, fallbackFn = this.getNavigationFallback()) {
        if (!path || path.length === 0) {
            return false
        }

        const destination = path[path.length - 1]
        if (path.length < 2) {
            return true
        }

        const target = { x: destination.x, y: destination.y }
        return this._enqueueResolvedPath(path, () => this.recoverNavigation(target, fallbackFn), true)
    }

    appendMovePath(path, fallbackFn = this.getNavigationFallback()) {
        if (!path || path.length === 0) {
            return false
        }

        const destination = path[path.length - 1]
        if (path.length < 2) {
            return true
        }

        const target = { x: destination.x, y: destination.y }
        return this._enqueueResolvedPath(path, () => this.recoverNavigation(target, fallbackFn), false)
    }

    getQueuedPathPreview() {
        if (this.pathPreview.length === 0) {
            return []
        }
        return [{ x: this.location.x, y: this.location.y }, ...this.pathPreview]
    }

    move() {
        let nextAction = this.queue.dequeue()

        if (nextAction === null && typeof this.getBehavior === 'function') {
            const behavior = this.getBehavior()
            if (typeof behavior === 'function') {
                behavior.call(this)
                nextAction = this.queue.dequeue()
            }
        }

        if (nextAction === null) {
            return null
        }
        return nextAction()
    }

    getSpritePlacementForTile(tileSprite, { randomize = false } = {}) {
        if (!tileSprite || !Number.isFinite(tileSprite.x) || !Number.isFinite(tileSprite.y)) {
            return null
        }

        const scale = Number.isFinite(this.game?.currentScale) ? this.game.currentScale : 1
        const baseX = Number.isFinite(tileSprite.baseX) ? tileSprite.baseX : tileSprite.x
        const baseY = Number.isFinite(tileSprite.baseY) ? tileSprite.baseY : tileSprite.y
        const offset = randomize ? getRandomMovementOffset() : { x: 0, y: 0 }

        return {
            x: tileSprite.x + (offset.x * scale),
            y: tileSprite.y + (offset.y * scale),
            baseX: baseX + offset.x,
            baseY: baseY + offset.y
        }
    }

    placeSpriteOnTile(tileSprite, options = {}) {
        const placement = this.getSpritePlacementForTile(tileSprite, options)
        if (!placement || !this.sprite) {
            return false
        }

        this.sprite.x = placement.x
        this.sprite.y = placement.y
        this.sprite.baseX = placement.baseX
        this.sprite.baseY = placement.baseY
        return true
    }

    performMove(next) {
        return this.cave.moveCreature(this, next)
    }

    getActions() {
        let myTile = this.cave.getTile(toKey(this.location))
        let myNeighbors = [ ...myTile.getNeighbors()]

        var myActions = new Set()

        if (myTile.getBuilt()) {
            let myBuilding = myTile.getBuilt()
            if (myBuilding.hasStation) {
                myActions.add(myBuilding)
            }
        } 

        for (let n of myNeighbors) {
            if (n.getBuilt()) {
                if (!n.getBuilt().hasStation) {
                    myActions.add(n.getBuilt())
                }
            }
        }

        return myActions
    }

    getBuildable() {
        let myBuildings = [...this.game.unlockedBuildings]
        //if creature can build special buildings, add those here
        return myBuildings
    }
}

File: TriloGame.CSharp/src/TriloGame.Game/Core/Entities/Creature.cs
using Microsoft.Xna.Framework;
using TriloGame.Game.Core.Buildings;
using TriloGame.Game.Core.Simulation;
using TriloGame.Game.Shared.Math;
using TriloGame.Game.Shared.Utilities;

namespace TriloGame.Game.Core.Entities;

public class Creature
{
    private const float MovementOffsetMinDistance = 1f;
    private const float MovementOffsetMaxDistance = 15f;
    private readonly Queue<Action> _queue = new();

    public Creature(string name, GridPoint location, GameSession session)
    {
        Name = name;
        Location = location;
        Session = session;
        Health = 20;
        MaxHealth = 20;
        Damage = 5;
        Assignment = "unassigned";
        MovementOffset = Vector2.Zero;
        RotationRadians = 0f;
        PathPreview = [];
    }

    public string Name { get; }

    public List<GridPoint> PathPreview { get; }

    public int Health { get; protected set; }

    public int MaxHealth { get; protected set; }

    public int Damage { get; protected set; }

    public GridPoint Location { get; set; }

    public float RotationRadians { get; set; }

    public Vector2 MovementOffset { get; private set; }

    public GameSession Session { get; }

    public World.Cave? Cave { get; set; }

    public string Assignment { get; set; }

    public void ClearActionQueue()
    {
        _queue.Clear();
        PathPreview.Clear();
    }

    public bool RestartBehavior(bool clearQueue = true)
    {
        if (clearQueue)
        {
            ClearActionQueue();
        }

        var behavior = GetBehavior();
        if (behavior is null)
        {
            return false;
        }

        behavior();
        return true;
    }

    public virtual Action? GetBehavior() => null;

    public int RestoreHealth()
    {
        Health = MaxHealth;
        return Health;
    }

    public int DealDamage(object? target)
    {
        return target switch
        {
            Creature creature when !ReferenceEquals(creature, this) => creature.TakeDamage(Damage, this),
            Building building => building.TakeDamage(Damage, this),
            _ => 0
        };
    }

    public virtual int TakeDamage(int amount, object? source = null)
    {
        if (amount <= 0 || Health <= 0)
        {
            return 0;
        }

        var applied = System.Math.Min(Health, amount);
        Health -= applied;
        if (Health <= 0)
        {
            Health = 0;
            RemoveFromGame(source);
        }

        return applied;
    }

    public virtual void CleanupBeforeRemoval(object? source = null)
    {
    }

    public virtual bool RemoveFromGame(object? source = null)
    {
        return Cave?.RemoveCreature(this, source) ?? true;
    }

    public bool EnqueueAction(Action action)
    {
        _queue.Enqueue(action);
        return true;
    }

    public virtual Action? GetNavigationFallback()
    {
        return Assignment switch
        {
            "miner" when this is Trilobite trilobite => () => trilobite.MinerStep1(),
            "farmer" when this is Trilobite trilobite => () => trilobite.FarmerStep1(),
            "builder" when this is Trilobite trilobite => () => trilobite.BuilderStep1(),
            "fighter" when this is Trilobite trilobite => () => trilobite.FighterStep1(),
            "enemy" when this is Enemy enemy => () => enemy.EnemyStep1(),
            _ => null
        };
    }

    protected bool RunNavigationFallback(Action? fallbackFn)
    {
        ClearActionQueue();
        if (fallbackFn is not null)
        {
            EnqueueAction(fallbackFn);
        }

        return false;
    }

    public List<GridPoint>? BuildNavigationPathToPoint(GridPoint destination)
    {
        return Cave?.BuildPathFromField(Cave.BuildPointBfsField(destination), Location);
    }

    public List<GridPoint>? BuildNavigationPathToBuilding(Building building)
    {
        return Cave?.BuildPathFromField(Cave.EnsureBuildingBfsField(building), Location);
    }

    protected bool RecoverNavigation(GridPoint? destination, Action? fallbackFn)
    {
        ClearActionQueue();
        if (destination is null)
        {
            return RunNavigationFallback(fallbackFn);
        }

        var reroute = BuildNavigationPathToPoint(destination.Value);
        if (reroute is not null && reroute.Count > 1)
        {
            EnqueueResolvedPath(reroute, () => RecoverNavigation(destination, fallbackFn), false);
            return false;
        }

        return reroute is not null && reroute.Count == 1 || RunNavigationFallback(fallbackFn);
    }

    protected bool RecoverBuildingNavigation(Building? building, Action? fallbackFn)
    {
        ClearActionQueue();
        if (building is null)
        {
            return RunNavigationFallback(fallbackFn);
        }

        var reroute = BuildNavigationPathToBuilding(building);
        if (reroute is not null && reroute.Count > 1)
        {
            EnqueueResolvedPath(reroute, () => RecoverBuildingNavigation(building, fallbackFn), false);
            return false;
        }

        return reroute is not null && reroute.Count == 1 || RunNavigationFallback(fallbackFn);
    }

    protected bool ExecuteNavigationStep(GridPoint next, Action? onFailure)
    {
        var result = Cave?.MoveCreature(this, next);
        if (result == false)
        {
            onFailure?.Invoke();
            return false;
        }

        if (PathPreview.Count > 0)
        {
            PathPreview.RemoveAt(0);
        }

        return result ?? false;
    }

    protected bool EnqueueResolvedPath(IReadOnlyList<GridPoint> path, Action? onFailure, bool clearExisting)
    {
        if (clearExisting)
        {
            ClearActionQueue();
        }

        if (path.Count < 2)
        {
            return false;
        }

        foreach (var step in path.Skip(1))
        {
            var next = step;
            PathPreview.Add(next);
            EnqueueAction(() => ExecuteNavigationStep(next, onFailure));
        }

        return true;
    }

    public bool NavigateTo(GridPoint destination, Action? fallbackFn = null, bool clearExisting = true)
    {
        fallbackFn ??= GetNavigationFallback();
        var path = BuildNavigationPathToPoint(destination);
        if (path is null)
        {
            return RunNavigationFallback(fallbackFn);
        }

        return path.Count < 2 || EnqueueResolvedPath(path, () => RecoverNavigation(destination, fallbackFn), clearExisting);
    }

    public bool NavigateToBuilding(Building building, Action? fallbackFn = null, bool clearExisting = true)
    {
        fallbackFn ??= GetNavigationFallback();
        var path = BuildNavigationPathToBuilding(building);
        if (path is null)
        {
            return RunNavigationFallback(fallbackFn);
        }

        return path.Count < 2 || EnqueueResolvedPath(path, () => RecoverBuildingNavigation(building, fallbackFn), clearExisting);
    }

    public bool QueueMovePath(IReadOnlyList<GridPoint> path, Action? fallbackFn = null)
    {
        fallbackFn ??= GetNavigationFallback();
        if (path.Count < 2)
        {
            return path.Count > 0;
        }

        var destination = path[^1];
        return EnqueueResolvedPath(path, () => RecoverNavigation(destination, fallbackFn), true);
    }

    public bool AppendMovePath(IReadOnlyList<GridPoint> path, Action? fallbackFn = null)
    {
        fallbackFn ??= GetNavigationFallback();
        if (path.Count < 2)
        {
            return path.Count > 0;
        }

        var destination = path[^1];
        return EnqueueResolvedPath(path, () => RecoverNavigation(destination, fallbackFn), false);
    }

    public List<GridPoint> GetQueuedPathPreview()
    {
        return PathPreview.Count == 0 ? [] : [Location, .. PathPreview];
    }

    public object? Move()
    {
        if (_queue.Count == 0)
        {
            GetBehavior()?.Invoke();
        }

        if (_queue.Count == 0)
        {
            return null;
        }

        var action = _queue.Dequeue();
        action();
        return true;
    }

    public void UpdateMovementOffset(bool randomize)
    {
        MovementOffset = randomize
            ? RandomUtil.NextMovementOffset(MovementOffsetMinDistance, MovementOffsetMaxDistance)
            : Vector2.Zero;
    }

    public bool PerformMove(GridPoint next)
    {
        return Cave?.MoveCreature(this, next) ?? false;
    }

    public HashSet<Building> GetActions()
    {
        var currentTile = Cave?.GetTile(Location.ToString());
        if (currentTile is null)
        {
            return [];
        }

        var actions = new HashSet<Building>();
        if (currentTile.Built is { HasStation: true } currentBuilding)
        {
            actions.Add(currentBuilding);
        }

        foreach (var neighbor in currentTile.Neighbors)
        {
            if (neighbor.Built is { HasStation: false } building)
            {
                actions.Add(building);
            }
        }

        return actions;
    }

    public virtual List<Factory> GetBuildable()
    {
        return [.. Session.UnlockedBuildings];
    }
}
